lib/Basics/expected.h

```diff
@@ -0,0 +1,215 @@
+    new (&_value) T(std::forward<Args>(args)...);
+    _state = kValue;
+  }
+
+  explicit expected(std::exception_ptr ex) noexcept
+      : _exception(std::move(ex)), _state(kException) {
+    static_assert(std::is_nothrow_move_constructible_v<std::exception_ptr>);
+  }
+
+  expected(expected const& other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>) requires
+      std::is_copy_constructible_v<T> {
+    if (other._state == kValue) {
+      new (&_value) T(other._value);
+      _state = kValue;
+    } else if (other._state == kException) {
+      static_assert(std::is_nothrow_copy_constructible_v<std::exception_ptr>);
+      new (&_exception) std::exception_ptr(other._exception);
+      _state = kException;
+    }
+  }
+
+  expected(expected&& other) noexcept(
+      std::is_nothrow_move_constructible_v<T>) requires
+      std::move_constructible<T> {
+    if (other._state == kValue) {
+      new (&_value) T(std::move(other._value));
+      _state = kValue;
+    } else if (other._state == kException) {
+      static_assert(std::is_nothrow_move_constructible_v<std::exception_ptr>);
+      new (&_exception) std::exception_ptr(std::move(other._exception));
+      _state = kException;
+    }
+  }
+
+  expected& operator=(expected const& other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>&&
+          std::is_nothrow_destructible_v<T>&& std::is_nothrow_copy_assignable_v<
+              T>) requires(std::is_copy_constructible_v<T>&&
+                               std::is_copy_assignable_v<T>) {
+    if (this != &other) {
+      if (other._state == kValue && _state == kValue) {
+        _value = other._value;
+      } else {
+        reset();
+        if (other._state == kValue) {
+          new (&_value) T(other._value);
+          _state = kValue;
+        } else if (other._state == kException) {
+          static_assert(
+              std::is_nothrow_copy_constructible_v<std::exception_ptr>);
+          new (&_exception) std::exception_ptr(other._exception);
+          _state = kException;
+        }
+      }
+    }
+    return *this;
+  }
+
+  expected& operator=(expected&& other) noexcept(
+      std::is_nothrow_move_constructible_v<T>&&
+          std::is_nothrow_destructible_v<T>&& std::is_nothrow_move_assignable_v<
+              T>) requires(std::move_constructible<T>&&
+                               std::is_move_assignable_v<T>) {
+    if (this != &other) {
+      if (other._state == kValue && _state == kValue) {
+        _value = std::move(other._value);
+      } else {
+        reset();
+        if (other._state == kValue) {
+          new (&_value) T(std::move(other._value));
+          _state = kValue;
+        } else if (other._state == kException) {
+          static_assert(
+              std::is_nothrow_move_constructible_v<std::exception_ptr>);
+          new (&_exception) std::exception_ptr(std::move(other._exception));
+          _state = kException;
+        }
+        other.reset();
```

Comment:
Same code as in the move constructor - can we unify this? (Here we have the reset that is missing in the move constructor. Yet another reason for unification ;)